### PR TITLE
[IOTDB-5765] Support ORDER BY expression

### PR DESCRIPTION
**Pull request by ycycse**
userGuide: https://apache-iotdb.feishu.cn/docx/Cqp6dhiEVoUakGx0wg1clGGlneg
design doc: https://apache-iotdb.feishu.cn/docx/T1msdaV6OoTLczxzQmQccdYunif

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/merge/MergeSortKeyComparator.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  @Override
+  public Comparator<MergeSortKey> thenComparing(Comparator<? super MergeSortKey> other) {
+    Objects.requireNonNull(other);
+    return new MergeSortKeyComparator(index, nullFirst, originalComparator.thenComparing(other));
+  }
+
+  @Override
+  public Comparator<MergeSortKey> reversed() {
+    return new MergeSortKeyComparator(index, !nullFirst, originalComparator.reversed());
+  }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/merge/MergeSortKeyComparator.java:77]: It seems that these two methods are not used in your implementation.

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/merge/MergeSortKeyComparator.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    this.nullFirst = nullFirst;
+    this.index = index;
+    this.originalComparator = originalComparator;
+  }
+
+  public static MergeSortKeyComparator getInstance(
+      int index, boolean nullFirst, Comparator<MergeSortKey> originalComparator) {
+    return comparatorCache.computeIfAbsent(
+        new Pair<>(originalComparator, new Pair<>(nullFirst, index)),
+        comparator -> new MergeSortKeyComparator(index, nullFirst, originalComparator));
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/merge/MergeSortKeyComparator.java:51]: This cache doesn't take effective at all. Each time the originalComparator won't be the same.

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanBuilder.java [hunk 3]
```diff
@@ -1200,4 +1197,36 @@ private LogicalPlanBuilder planSingleShowQueries(TDataNodeLocation dataNodeLocat
...
+      QueryStatement queryStatement,
+      Set<Expression> orderByExpressions,
+      Set<Expression> selectExpression) {
+    // only the order by clause having expression needs a sortNode
+    if (!queryStatement.hasOrderByExpression()) {
+      return this;
+    }
+
+    List<SortItem> sortItems = queryStatement.getSortItemList();
+    queryStatement.updateSortItems(orderByExpressions);
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LogicalPlanBuilder.java:1211]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/merge/MergeSortKeyComparator.java (hunk 1 above)
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/merge/MergeSortKeyComparator.java:77]: done

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**